src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,714 @@
+                        log_syntax(unit, LOG_ERR, filename, line, r, "Failed to split allowed ips \"%s\" option: %m", rvalue);
+                        break;
+                }
+
+                r = in_addr_prefix_from_string_auto(word, &family, &addr, &prefixlen);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r, "Network address is invalid, ignoring assignment: %s", word);
+                        return 0;
+                }
+
+                ipmask = new0(WireguardIpmask, 1);
+                if (!ipmask)
+                        return log_oom();
+                ipmask->family = family;
+                ipmask->ip6 = addr.in6;
+                ipmask->cidr = prefixlen;
+
+                LIST_PREPEND(ipmasks, peer->ipmasks, ipmask);
+        }
+
+        return 0;
+}
+
+int config_parse_wireguard_endpoint(const char *unit,
+                                    const char *filename,
+                                    unsigned line,
+                                    const char *section,
+                                    unsigned section_line,
+                                    const char *lvalue,
+                                    int ltype,
+                                    const char *rvalue,
+                                    void *data,
+                                    void *userdata) {
+        Wireguard *w;
+        WireguardPeer *peer;
+        size_t len;
+        const char *begin, *end = NULL;
+        _cleanup_free_ char *host = NULL, *port = NULL;
+        _cleanup_free_ WireguardEndpoint *endpoint = NULL;
+
+        assert(data);
+        assert(rvalue);
+
+        w = WIREGUARD(data);
+
+        assert(w);
+
+        peer = wireguard_peer_new(w, section_line);
+        if (!peer)
+                return log_oom();
+
+        endpoint = new0(WireguardEndpoint, 1);
+        if (!endpoint)
+                return log_oom();
+
+        if (rvalue[0] == '[') {
+                begin = &rvalue[1];
+                end = strchr(rvalue, ']');
+                if (!end) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0, "Unable to find matching brace of endpoint, ignoring assignment: %s", rvalue);
+                        return 0;
+                }
+                len = end - begin;
+                ++end;
+                if (*end != ':' || !*(end + 1)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0, "Unable to find port of endpoint: %s", rvalue);
+                        return 0;
+                }
+                ++end;
+        } else {
+                begin = rvalue;
+                end = strrchr(rvalue, ':');
+                if (!end || !*(end + 1)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0, "Unable to find port of endpoint: %s", rvalue);
+                        return 0;
+                }
+                len = end - begin;
+                ++end;
+        }
```

Comment:
maybe use socket_address_parse() for this? or some IP-specific part of it, split into a new function socket_address_parse_ip() or so?

Reply:
this function probably did not have existed, when I originally wrote it.

Reply:
Is `socket_address_parse` flexible enough? Using it certainly seems nice, but something to keep in mind is that an Endpoint is allowed to be a DNS name. These are all valid endpoints:

- demo.wireguard.com:1234
- [abcd::1]:2468
- [fe80:efef::1%eth0]:2468
- 1.2.3.4:9876
- 1.2.3.4:http

What `wg(8)` does is just pass the host part in the host param and the port part in the port param to getaddrinfo and then have getaddrinfo work out how to resolve the differing parts. It then returns a sockaddr. Since getaddrinfo is doing all the heavy lifting, all we really need is a function to split apart the parts of the endpoint. It turns out this is basically what everybody who accepts hostnames does too -- think about URLs.

So, can `socket_address_parse` do this? I suspect not, and we'd actually need an asynchronous `socket_endpoint_parse`, which does the right thing w.r.t. asynchronous systemd's resolver APIs. This would be, in other words, an abstraction of that logic that @Mic92 has already implemented in here. Abstractions are sometimes nice, sometimes not nice. What would you think of keeping this as is, and abstracting it only when somebody else needs that functionality, should that day come? Or would you prefer to abstract it out early on?

Reply:
The example should all work except service names in ports, where @poettering had a opinion against: https://github.com/systemd/systemd/pull/4191#discussion_r87632131

Reply:
Ahh, right:

> service names are stupid. I really wouldn't bother. service names never worked, nobody uses them, and they should just die..
 
Okay, no problem. They're not especially useful in this context either (where people enjoy putting wireguard on all sorts of weird ports, not just 51820), so if @poettering isn't up for it, let's roll with it.

Reply:
You were also right, that `socket_address_parse` does not support hostnames.